# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmFrameIter.cpp
# Commit: 1f21e8f610a7
# Full Hash: 1f21e8f610a72d6c6bd84321f41fb257335213e4
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-10-31 21:05:17
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 5: Remove remaining uses of Frame::tls. r=lth
#   
#   Here we remove remaining uses of Frame::tls. There are many places where
#   we use it: in debug frames, in profiling frames, in jit activation, etc.
#   All these places require short fixes to use our new scheme for getting
# ==============================================================================

diff -r c50443ca5619 -r 1f21e8f610a7 js/src/wasm/WasmFrameIter.cpp
--- a/js/src/wasm/WasmFrameIter.cpp	Sat Oct 31 09:56:05 2020 +0000
+++ b/js/src/wasm/WasmFrameIter.cpp	Sat Oct 31 09:56:05 2020 +0000
@@ -207,9 +207,7 @@
     tls_ = ExtractCallerTlsFromFrameWithTls(prevFP);
   }
 
-  MOZ_ASSERT(code_ == &fp_->instance()->code());
-  MOZ_ASSERT(fp_->instance() == tls()->instance);
-
+  MOZ_ASSERT(code_ == &tls()->instance->code());
   lineOrBytecode_ = callsite->lineOrBytecode();
 
   MOZ_ASSERT(!done());
@@ -928,7 +926,7 @@
          (offsetInCode - codeRange->funcCheckedCallEntry()) > SetFP;
 }
 
-TlsData* js::wasm::GetNearestEffectiveTls(Frame* fp) {
+const TlsData* js::wasm::GetNearestEffectiveTls(const Frame* fp) {
   while (true) {
     if (fp->callerIsExitOrJitEntryFP()) {
       // It is a direct call from JIT.
@@ -956,6 +954,11 @@
   }
 }
 
+TlsData* js::wasm::GetNearestEffectiveTls(Frame* fp) {
+  return const_cast<TlsData*>(
+      GetNearestEffectiveTls(const_cast<const Frame*>(fp)));
+}
+
 bool js::wasm::StartUnwinding(const RegisterState& registers,
                               UnwindState* unwindState, bool* unwoundCaller) {
   // Shorthands.
@@ -1327,7 +1330,8 @@
   }
 
   MOZ_ASSERT(code_ ==
-             &Frame::fromUntaggedWasmExitFP(callerFP_)->instance()->code());
+             &GetNearestEffectiveTls(Frame::fromUntaggedWasmExitFP(callerFP_))
+                  ->instance->code());
 
   switch (codeRange_->kind()) {
     case CodeRange::Function: